### PR TITLE
[clang][APINotes] Add support for the SwiftEscapable attribute

### PR DESCRIPTION
**Pull request by Xazax-hun**
This is similar to SwiftCopyable. Also fix missing SwiftCopyable dump for TagInfo.

### CODE REVIEW

File: clang/lib/APINotes/APINotesWriter.cpp [hunk 1]
```diff
@@ -1266,11 +1266,11 @@ class CommonTypeTableInfo
...
-    return 2 + (TI.SwiftImportAs ? TI.SwiftImportAs->size() : 0) +
-           2 + (TI.SwiftRetainOp ? TI.SwiftRetainOp->size() : 0) +
-           2 + (TI.SwiftReleaseOp ? TI.SwiftReleaseOp->size() : 0) +
-           2 + (TI.SwiftConformance ? TI.SwiftConformance->size() : 0) +
-           2 + getCommonTypeInfoSize(TI);
+    return 2 + (TI.SwiftImportAs ? TI.SwiftImportAs->size() : 0) + 2 +
+           (TI.SwiftRetainOp ? TI.SwiftRetainOp->size() : 0) + 2 +
+           (TI.SwiftReleaseOp ? TI.SwiftReleaseOp->size() : 0) + 2 +
+           (TI.SwiftConformance ? TI.SwiftConformance->size() : 0) + 3 +
+           getCommonTypeInfoSize(TI);
```
Comment (egorzhdan) [inline clang/lib/APINotes/APINotesWriter.cpp:1273]: Last time I touched this code, I think @compnerd preferred to not clang-format this, but I don't have a super strong opinion on this.

**Review by egorzhdan — APPROVED** [review]
LGTM

File: clang/include/clang/APINotes/Types.h [hunk 2]
```diff
@@ -757,6 +763,15 @@ class TagInfo : public CommonTypeInfo {
     SwiftCopyable = Value.value_or(false);
   }
 
+  std::optional<bool> isSwiftEscapable() const {
+    return SwiftEscapableSpecified ? std::optional<bool>(SwiftEscapable)
+                                   : std::nullopt;
+  }
```
Comment (compnerd) [inline clang/include/clang/APINotes/Types.h:769]:
```suggestion
  }

```

File: clang/lib/APINotes/APINotesWriter.cpp (hunk 1 above)
Comment (compnerd) [reply clang/lib/APINotes/APINotesWriter.cpp:1273]: Yeah, I think that the grouping helps understand what is being added up (aka, don't clang-format this).

File: clang/lib/APINotes/APINotesReader.cpp [hunk 3]
```diff
@@ -589,10 +589,16 @@ class TagTableInfo
...
       Info.setSwiftCopyable(std::optional(false));
-    else if (Copyable == kSwiftCopyable)
+    else if (Copyable == kSwiftConforms)
       Info.setSwiftCopyable(std::optional(true));
+    uint8_t Escapable =
+        endian::readNext<uint8_t, llvm::endianness::little>(Data);
+    if (Escapable == kSwiftDoesNotConform)
+      Info.setSwiftEscapable(std::optional(false));
+    else if (Escapable == kSwiftConforms)
+      Info.setSwiftEscapable(std::optional(true));
```
Comment (compnerd) [inline clang/lib/APINotes/APINotesReader.cpp:601]:
```suggestion
    if (Escapable == kSwiftConforms || Escapable == kSwiftDoesNotConform)
      Info.setSwiftEscapable(std::optional(Escapable == kSwiftConforms));
```

File: clang/lib/Sema/SemaAPINotes.cpp [hunk 4]
```diff
@@ -647,6 +647,13 @@ static void ProcessAPINotes(Sema &S, TagDecl *D, const api_notes::TagInfo &Info,
       D->addAttr(SwiftAttrAttr::Create(S.Context, "~Copyable"));
   }
 
+  if (auto Escapable = Info.isSwiftEscapable()) {
+    if (*Escapable)
+      D->addAttr(SwiftAttrAttr::Create(S.Context, "Escapable"));
+    else
+      D->addAttr(SwiftAttrAttr::Create(S.Context, "~Escapable"));
+  }
```
Comment (compnerd) [inline clang/lib/Sema/SemaAPINotes.cpp:655]:
```suggestion
  if (auto Escapable = Info.isSwiftEscapable())
    D->addAttr(SwiftAttrAttr::Create(S.Context, *Escapable ? "Escapable" : "~Escapable"));
```

**Review by compnerd — APPROVED** [review]
Some minor comments left.

File: clang/lib/APINotes/APINotesWriter.cpp (hunk 1 above)
Comment (Xazax-hun) [reply clang/lib/APINotes/APINotesWriter.cpp]: I added clang format directives to prevent this from happening in the future.

**Pull request merged**